### PR TITLE
Add score in practice quiz

### LINKED ISSUE #8643: Update score on content card for practice quizzes
## Observed behavior
The content cards currently do not have the ability to show the score. We will need to connect to the backend to pull information for the most recent score for a practice quiz from the learner to populate on the cards.

## Expected behavior
[See the Figma here for the practice quiz cards.](https://www.figma.com/file/TfaSTT9gs6fPKhGdNUjffU/Past-papers-project?node-id=911%3A5165)

Screen Shot 2021-11-09 at 2 11 46 PM

## User-facing consequences
Currently, users will see that a practice quiz has been completed, but will not have the ability to see their most recent score until they go into the learner report.

### PR DESCRIPTION
**Pull request by thesujai**
## Summary

1. Update the ContentNodeProgressViewset to return num_question_answered, num_question_answered_correctly, and total_questions
2. New map for the only quizprogress separated from other content progress
3. ProgressBar edited to deal with practice quizzes separately
4. New corestring label for questions left

**Screenshot**
Screenshot from 2024-09-05 14-29-14

## References

Fixes #8643 

## Reviewer guidance

Please suggest if the reactivity needs to be improved

----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by MisRob** [comment]
Hey @thesujai, thanks! Regarding frontend parts,

> How do I calculate the score and no of question in the frontend

Was this a question for us? If so, we have some score calculations here 

https://github.com/learningequality/kolibri/blob/develop/kolibri/plugins/learn/assets/src/views/cards/QuizCard/index.vue#L59-L87

That said, I'm not entirely sure what cards exactly you will need work on. Please send me a link to the card component that's used for a practice quiz.

**Comment by MisRob** [comment]
Also there's some guidance in comments from @rtibbles in the issue's comments https://github.com/learningequality/kolibri/issues/8643, and some touch on frontend parts as well.

**Comment by thesujai** [comment]
HI @MisRob 
The issue was that we could keep attempting a practice quiz even after all questions were complete(Not a bug).  But this means that a new attempt_log will be created for every new attempt. So a point can be reached where `num_correct_attempts>num_questions`. We cannot calculate score in this case.
I discussed this with @nucleogenesis and he suggested fetching all of the attemptlogs which are correct, but unique by AttemptLog.item. So I am working on that approach

**Comment by MisRob** [comment]
Okay, it sounds you've got all you need then

**Comment by thesujai** [comment]
Concern:
When a quiz has more questions than the minimum required to complete it (e.g., a quiz has 20 questions, but only 6 need to be correct to mark it as complete), the progress shows as 1 (100%) after answering the required number of questions. However, if the score is calculated using the total number of questions (num_correct_attempts/num_questions), it may appear low even when the quiz is marked as complete, which could be confusing.

One approach could be to calculate num_questions and num_correct_attempts based on the criteria for completing the quiz. For instance, if a quiz is marked complete after 6 correct answers, then:

- `num_questions` would be set to 6 (the number required for completion).
- `num_correct_attempts` would be calculated up to that point, ensuring the score reflects only those necessary attempts.

This way, the score and progress would be more aligned, avoiding confusion when the progress is 1 but the score seems low.

**Comment by MisRob** [comment]
thesujai I am looping in @nucleogenesis here, since you mentioned you've already had some discussions regarding this together

**Comment by MisRob** [comment]
@thesujai I am looping in @nucleogenesis here, since you mentioned you've already had some discussions regarding this together

**Review by nucleogenesis — COMMENTED** [review]
@thesujai the query code looks good, but I'm honestly not entirely sure how best to display this information to the user.

So I think that what we need is a design decision (@jtamiace @rtibbles ) w/ regard to what we actually want to express to the user here.

We have three things:
- Total number of questions in the Practice quiz `total_questions`
- Total time the user has attempted any question at all (correct or incorrect) `num_attempts`
- The total number of questions the user has answered correctly `num_correct_attempts` (not the number of _times_ questions are answered correctly, but the number of questions which have ever been answered correctly)

Example (not super realistic, but for demonstration/ensuring I'm understanding):
- Practice quiz with 5 questions `total_questions = 5`
- User answers all of those questions incorrectly 3 times each `num_attempts = 15`
- User answers the **same question** correctly 5 times `num_correct_attempts=1; num_attempts=20;`
- User answers the rest of the questions correctly in a row `num_correct_attempts=5; num_attempts=24`

So with these values... what do we want to tell the user? Are these the values we expect to use here or have I been misleading @thesujai in our discussions 😅

**Comment by nucleogenesis** [comment]
@thesujai after a quick chat w/ Richard I think I see the misunderstanding on my end.

A "Practice Quiz" will not allow multiple attempts at the same question within the same "Quiz Attempt" -- that is to say, when you take a "Practice Quiz" it is treated like a "Quiz" in that you're going straight through it.

Which means:
- Mastery model can be ignored
- % is a calculation of num_correct_attempts / total_questions because the user is expected to have attempted all of the questions exactly one time

So if you only get the data for the last "AttemptLog" for the practice quiz, then you should have the correct data there.

I'll give the code another look through w/ this in mind

**Comment by thesujai** [comment]
I understand why the linting failed here, but not why locally it passed

**Review by nucleogenesis — COMMENTED** [review]
@thesujai this is great work. I see my practice quiz's most recent score and things worked as expected.

I found one non-blocking issue which I will make an issue for once we merge this PR. You've put so much work into this and I think this is good to merge. I'll leave final review to @rtibbles next week on the code side to get another set of eyes on the queries and such.

---

For posterity, once this is merged, we'll need to make an issue to address this:

When you complete an attempt at a practice quiz and go back to the page listing the contents, the old score is shown instead of the new score.

I gave it a quick look and couldn't quickly work out a good way to have it basically re-load the progress but I suspect there is a way.

File: kolibri/core/assets/src/mixins/commonCoreStrings.js
Comment (nucleogenesis) [inline kolibri/core/assets/src/mixins/commonCoreStrings.js]: Ah - one blocking note here. I missed this was part of the diff in my initial review and thought it was pre-existing when I saw this.

We'll need to use ICU syntax for the word "questions" here which allows translators to translate the various ways of expressing an amount of something. A similar string is defined [here in the `QuizCard` component](https://github.com/learningequality/kolibri/blob/ec915139ef840970e56d287d4dafc88937eda190/kolibri/plugins/learn/assets/src/views/cards/QuizCard/index.vue#L90-L94) which demonstrates how this works where we can say "when the value of `questionsLeft` is `one` then the word is `question` otherwise, use `questions`".

You could probably just copy that string directly into the `ProgressBar` component - note how you'll add a `$tr: {}` object to that component, then add the string inside of that object there.

**Review by nucleogenesis — DISMISSED** [review]
Noticed one thing w/ the string added

File: kolibri/core/assets/src/mixins/commonCoreStrings.js
Comment (thesujai) [reply kolibri/core/assets/src/mixins/commonCoreStrings.js]: Interesting! ✅

File: kolibri/plugins/learn/assets/src/views/ProgressBar.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/ProgressBar.vue]: Might be good to specifically label this as `quizInProgressLabel` so that it's clear that it is only being used in that circumstance.

File: kolibri/plugins/learn/assets/src/views/ProgressBar.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/ProgressBar.vue]: Likewise this could be `quizCompletedLabel`. Otherwise it's slightly confusing why there is `completedLabel` in the template above and then `coreString('completedLabel')` below it.

File: kolibri/plugins/learn/assets/src/views/ProgressBar.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/ProgressBar.vue:12]: This will add an extraneous `<p>` tag in the case that it this is not a quiz and not completed. We should retain the `v-if` but change it to:
```
v-if="completed || isQuiz"
```

File: kolibri/plugins/learn/assets/src/views/ProgressBar.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/ProgressBar.vue]: With the above tweak to the `v-if` this can just be a `v-else` instead.

**Review by rtibbles — DISMISSED** [review]
A couple of suggestions - ensuring we don't have an extraneous `<p>` is blocking, in case it does unexpected things to layout.

File: kolibri/plugins/learn/assets/src/views/ProgressBar.vue
Comment (thesujai) [reply kolibri/plugins/learn/assets/src/views/ProgressBar.vue:12]: I didn't notice this one

File: kolibri/plugins/learn/assets/src/views/ProgressBar.vue
Comment (thesujai) [reply kolibri/plugins/learn/assets/src/views/ProgressBar.vue]: I added a ternary operator to display score or inProgress in a quiz, this eliminated the use of one more `else-if` statement, not functionally but readability-wise.

**Review by rtibbles — COMMENTED** [review]
Very neatly done!

**Comment by rtibbles** [comment]
Codewise, this is good to go - this just needs QA sign off to merge.

**Comment by pcenov** [comment]
Hi @rtibbles and @thesujai besides the issue already noted in https://github.com/learningequality/kolibri/issues/12663 I was able to identify only one more which is that the 'Score N%' text is not translated in the other supported languages:

score

**Comment by rtibbles** [comment]
As this is new string here, I think this is to be expected!

**Review by rtibbles — APPROVED** [review]
No QA concerns, string will be translated at the next translation round.

**Pull request merged**
